pandas/tests/reshape/merge/test_merge.py

```diff
@@ -1526,6 +1526,17 @@ def test_merge_on_ints_floats_warning(self):
             result = B.merge(A, left_on='Y', right_on='X')
             assert_frame_equal(result, expected[['Y', 'X']])
 
+    def test_merge_incompat_infer_object(self):
+        # GH21119
+        df1 = DataFrame({'key': Series([True, False], dtype=object)})
+        df2 = DataFrame({'key': [True, False]})
+
+        expected = DataFrame({'key': [True, False]}, dtype=object)
+        result = pd.merge(df1, df2, on='key')
```

Comment:
need to test the error case (e.g. bool / inferred bool) on 1 side and numeric on the other (I *think* this will raise), prob have an existing test, just need to add to it